lib/HLSL/DxilCondenseResources.cpp

```diff
@@ -2585,9 +2589,17 @@ static void MarkCBUsesForExtractElement(
     ExtractValueInst *EV, bool bMinPrecision) {
   unsigned lowHighWordUsage = 0;
   unsigned evOffset = GetOffsetForCBExtractValue(EV, bMinPrecision, lowHighWordUsage);
+
+  // For tbuffer, where value extracted is always 32-bits:
+  // If lowHighWordUsage is 0, it means 32-bits used.
+  // If field marked is < 32 bits, we still need to mark the high 16-bits as
+  // used, in case there is another 16-bit field.
+  // Since MarkCBUse could return 0 on non-basic type field, look for 16
+  // when determining whether we still need to mark high word as used.
+  bool highUnmarked = EV->getType()->getScalarSizeInBits() == 32;
   if (!lowHighWordUsage || 0 != (lowHighWordUsage & kLowWordUsed))
-    MarkCBUse(offset + evOffset, fieldMap);
-  if (lowHighWordUsage & kHighWordUsed)
+    highUnmarked &= MarkCBUse(offset + evOffset, fieldMap) == 16;
+  if (highUnmarked && (!lowHighWordUsage || 0 != (lowHighWordUsage & kHighWordUsed)))
```

Comment:
For !lowHighWordUsage || 0 != (lowHighWordUsage & kHighWordUsed)), could it just be !lowHighWordUsage?

Reply:
They're not equivalent.